src/client_side_request.cc

```diff
@@ -1404,53 +1404,48 @@ ClientRequestContext::sslBumpAccessCheck()
         return false;
     }
 
-    const Ssl::BumpMode bumpMode = http->getConn()->sslBumpMode;
-    if (http->request->flags.forceTunnel) {
-        debugs(85, 5, "not needed; already decided to tunnel " << http->getConn());
-        if (bumpMode != Ssl::bumpEnd)
-            http->al->ssl.bumpMode = bumpMode; // inherited from bumped connection
-        return false;
-    }
+    auto srvBump = http->getConn()->serverBump();
 
-    // If SSL connection tunneling or bumping decision has been made, obey it.
-    if (bumpMode != Ssl::bumpEnd) {
-        debugs(85, 5, HERE << "SslBump already decided (" << bumpMode <<
-               "), " << "ignoring ssl_bump for " << http->getConn());
-
-        // We need the following "if" for transparently bumped TLS connection,
-        // because in this case we are running ssl_bump access list before
-        // the doCallouts runs. It can be removed after the bug #4340 fixed.
-        // We do not want to proceed to bumping steps:
-        //  - if the TLS connection with the client is already established
-        //    because we are accepting normal HTTP requests on TLS port,
-        //    or because of the client-first bumping mode
-        //  - When the bumping is already started
-        if (!http->getConn()->switchedToHttps() &&
-                !http->getConn()->serverBump())
-            http->sslBumpNeed(bumpMode); // for processRequest() to bump if needed and not already bumped
-        http->al->ssl.bumpMode = bumpMode; // inherited from bumped connection
-        return false;
-    }
+    if (srvBump && srvBump->at(XactionStep::tlsBump2)) {
+        assert(http->request->method == Http::METHOD_CONNECT);
```

Comment:
Please discuss what this assert() means, and why we must check this condition during step2.

Reply:
Well, we "must" not check. 
Looks that I add it to debug bad states, eg where the bumping step is not updated correctly, eg when the ConnStateData->sslServerBump->step  remains at step2 after bump completed and the first tunneled request enters this code.

The ClientRequestContext::sslBumpAccessCheck is called for all HTTP requests enters the doCallouts. But if the current ConnStateData object marks bumping step2 means that this is a doCallouts for a CONNECT request at bumping step2. 

I did not remove it (yet).